src/AdvancedVI.jl

```diff
@@ -1,270 +1,149 @@
-
-Constructs the variational posterior from the `model` and performs the optimization
-following the configuration of the given `VariationalInference` instance.
-
-# Arguments
-- `model`: `Turing.Model` or `Function` z ↦ log p(x, z) where `x` denotes the observations
-- `alg`: the VI algorithm used
-- `q`: a `VariationalPosterior` for which it is assumed a specialized implementation of the variational objective used exists.
-- `getq`: function taking parameters `θ` as input and returns a `VariationalPosterior`
-- `θ`: only required if `getq` is used, in which case it is the initial parameters for the variational posterior
+    init(
+        rng::Random.AbstractRNG,
+        obj::AbstractVariationalObjective,
+        λ::AbstractVector,
+        restructure
+    )
+
+Initialize a state of the variational objective `obj` given the initial variational parameters `λ`.
+This function needs to be implemented only if `obj` is stateful.
+
+!!! warning
+    This is an internal function. Thus, the signature is subject to change without
+    notice.
 """
-function vi end
-
-function update end
-
-# default implementations
-function grad!(
-    vo,
-    alg::VariationalInference{<:ForwardDiffAD},
-    q,
-    model,
-    θ::AbstractVector{<:Real},
-    out::DiffResults.MutableDiffResult,
-    args...
-)
-    f(θ_) = if (q isa Distribution)
-        - vo(alg, update(q, θ_), model, args...)
-    else
-        - vo(alg, q(θ_), model, args...)
-    end
+init(
+    ::Random.AbstractRNG,
+    ::AbstractVariationalObjective,
+    ::AbstractVector,
+    ::Any
+) = nothing
 
-    # Set chunk size and do ForwardMode.
-    chunk_size = getchunksize(typeof(alg))
-    config = if chunk_size == 0
-        ForwardDiff.GradientConfig(f, θ)
-    else
-        ForwardDiff.GradientConfig(f, θ, ForwardDiff.Chunk(length(θ), chunk_size))
-    end
-    ForwardDiff.gradient!(out, f, θ, config)
-end
+"""
+    estimate_gradient!(
+        rng         ::Random.AbstractRNG,
+        prob,
+        adbackend   ::ADTypes.AbstractADType,
+        obj         ::AbstractVariationalObjective,
+        obj_state,
+        λ           ::AbstractVector,
+        restructure,
+        out         ::DiffResults.MutableDiffResult
+    )
+
+Estimate (possibly stochastic) gradients of the objective `obj` targeting `prob` with respect to the variational parameters `λ` using the automatic differentiation backend `adbackend`.
+The estimated objective value and gradient are then stored in `out`.
+If the objective is stateful, `obj_state` is its previous state, otherwise, it is `nothing`.
+
+# Returns
+- `out`: The `MutableDiffResult` containing the objective value and gradient estimates.
+- `obj_state`: The updated state of the objective estimator.
+- `stat`: Statistics and logs generated during estimation. (Type: `<: NamedTuple`)
+
+!!! warning
+    This is an internal function. Thus, the signature is subject to change without
+    notice.
```

Comment:
So this is not meant to be overloaded if I want to define my own VI method? It seems like it is, no? If so, we probably want to consider changes to this method breaking anyways, since otherwise it would break downstream implementations of VI methods.

Reply:
I.e. this warning shouldn't be necessary. Or am I misunderstanding the use of this method?

Reply:
Hmm, I see. I was only thinking "downstream" to be the internal use within the library, but I agree that there's no need to be that restrictive. I'll remove that warning.